moztelemetry/dataset.py

```diff
@@ -41,6 +42,39 @@ def _group_by_size_greedy(obj_list, tot_groups):
     return groups
 
 
+def _group_by_equal_size(obj_list, tot_groups, threshold=pow(2, 32)):
+    sub_list = []
+    groups = []
+    priority = []
+    tmp_object = None
+    sub_list_size = 0
+
+    if tot_groups <= 1:
+        groups = _group_by_size_greedy(obj_list, tot_groups)
+        return groups
+
+    for obj in obj_list:
+        for key in obj:
+            priority.append((obj[key], obj))
+
+    heapq.heapify(priority)
+    while priority:
+        tmp_object = heapq.heappop(priority)
+
+        if tmp_object[0] >= threshold:
+            groups.append(tmp_object)
+
+        sub_list.append(tmp_object)
+        sub_list_size += tmp_object[0]
+        if sub_list_size >= threshold:
+            groups.append(sub_list)
```

Comment:
I think the sub-lists should also be in a heap so that the next largest item gets added to the smallest list. The group should be a heap of length `tot_groups` e.g `[(0, []) for _ in range(tot_groups)]`.

Here's an example of where this approach won't pack optimally. Each row represents the successive sizes of the partitions. 

total groups = 3
threshold = 5
file sizes = 2, 2, 2

part 1| part 2| part 3
--|--|--
 0 | 0 | 0 
2 | 0 | 0
4 | 0 | 0
4 | 2 | 0

Instead, all of the partitions should be as evenly sized as possible. One possible way would be to add the next largest item to the emptiest partition:

part 1| part 2| part 3
--|--|--
 0 | 0 | 0 
2 | 0 | 0
2 | 2 | 0
2 | 2 | 2

Reply:
Okay, this makes sense. If this is the desired behavior, it does make more sense to use a second heap instead of a list. However, it seems I would then need to check the size of each group on every iteration in order to keep the groups equally sized. Is this true? Wanted to make sure I am on the right track with the desired implementation.

Reply:
Correct, there the size of the group needs to be checked on every iteration. If the size of the next item is added to the running group size, the heap can take care of the ordering though.